Sneak highlighting up to date (forked from #4718)

w-cantin opened the pull request:
Hello, 

This PR is a fork of the PR #4718 and is pretty much ready to merge. #4718 was stuck for a little while and I wanted to make sure it did not get abandoned. This PR is now properly rebased and up to date. The various fadeout issues mentioned in the previous PR are also now fixed. The escape key now properly escapes Sneak Label Input mode. When in Sneak Label Mode, the proper action is triggered if the key pressed does not correspond to any label (for example pressing y will trigger the yank action).

w-cantin commented:
It's ok @zhenbangbb . Those things take time. J-Fields is already doing a phenomenal job on VSCodeVim and all of this for free so no need to put even more pressure.

w-cantin commented:
Ok so I just finished refactoring a huge chunk of this PR. I wanted to reuse existing Easymotions functions more and also I kept finding new bugs with the old implementation. That is why a big part of sneak.ts has been rewritten. Hopefully the code is now easier to read and all the bugs are gone.

J-Fields commented:
Thanks so much for reviving and polishing this, @w-cantin! Sorry I've been so slow to get to your (and everyone else's) PRs - just been occupied with other things lately.

J-Fields commented on README.md:
```diff
@@ -561,10 +562,18 @@ Provided there is a new line between the opening and closing braces / tag, it ca
+| ---------------------------------- | -------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------- | ------- | ------------------------ |
+| vim.sneak                          | Enable/disable vim-sneak                                                                                                                                                                                                   | Boolean | false                    |
+| vim.sneakUseIgnorecaseAndSmartcase | Respect `vim.ignorecase` and `vim.smartcase` while sneaking                                                                                                                                                                | Boolean | false                    |
+| vim.sneakMaxLinesToConsider        | The maximum number of lines searched (starting from the position of the cursor) while sneaking (useful if one is working with large files). -1 means no restriction.                                                       | Number  | -1                       |
```
-1 strikes me as a slightly weird sentinel value. Any reason it's not 0 or `undefined` instead?

w-cantin replied:
I agree. The user could want to search only the current line though. At the moment maxLinesToConsider === 1 would mean to search the current line and the one below (if searching forward) and maxLinesToConsider === 0 would mean to search only the current line. I will put null as a default value as it makes more sense. Let me know if changing the behavior of 0 and 1 is required too.

J-Fields replied:
Maybe most natural to me is:

- Anything negative: invalid (I think you can set a minimum value via JSON schema)
- 0: all lines, default
- 1: just the current line
- 2: current line and next (or current line and previous)

Any issue with that scheme?

w-cantin replied:
None at all, I'll make the change soon.

J-Fields commented on package.json:
```diff
@@ -896,6 +927,14 @@
           "description": "Status bar color when in EasyMotionInput mode.",
           "default": "#007acc"
         },
+        "vim.statusBarColors.sneaklabelinputmode": {
```
`sneaklabelinputmode` should be camel case

w-cantin replied:
I am looking at the other settings for the status bar colors and they all follow the lowercase mode. Should I still change it to camelCase?

J-Fields commented on src/actions/plugins/sneak.ts:
```diff
@@ -1,138 +1,408 @@
+    return this.reorderMatches(matches);
+  }
+
+  protected searchWholeDocumennt(vimState: VimState, searchString: string): Match[] {
```
nit: typo in `Document`

J-Fields commented on README.md:
```diff
@@ -561,10 +562,18 @@ Provided there is a new line between the opening and closing braces / tag, it ca
-| vim.sneakUseIgnorecaseAndSmartcase | Respect `vim.ignorecase` and `vim.smartcase` while sneaking | Boolean | false         |
+| Setting                            | Description                                                                                                                                                                                                                | Type    | Default Value            |
+| ---------------------------------- | -------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------- | ------- | ------------------------ |
+| vim.sneak                          | Enable/disable vim-sneak                                                                                                                                                                                                   | Boolean | false                    |
```
Can you make sure these descriptions match what's in `package.json` please?

w-cantin commented:
Hello, 

I am not sure if this is the right place to ask, but there probably should be a way to stop the build server when the testing process is hanging. I inadvertently pushed some code that would hang during testing and there was no way for me to cancel the automatic github action. As of right now the testing server has been running for close to an hour unnecessarily. Maybe if there was a way to automatically fail the automatic build process after a certain amount of time it could prevent wasted cpu cycles.

PascalSenn commented:
There seems to be an error if you trie to search with F+"("

> Failed to handle key `(`: Invalid regular expression: /(/: Unterminated group

w-cantin commented:
Good eye! 

There was a bug when transforming searchString to a regex when the
searchString contained special Regex Characters like "(" or "%". It is now fixed and I added some tests to make sure it does not happen again.

w-cantin commented:
Hello @J-Fields,

I was wondering if adding label mode and highlighting to Sneak was still important for VSCodeVim. I am asking because I know it is a feature that many people want as seen by the number of downloads of a temporary fork I made [here](https://marketplace.visualstudio.com/items?itemName=vscodevim-sneak-fork.vscodevim-sneak-fork) and also the number of votes on this PR as well as the previous PR #4718.

I understand that you are quite busy and working on merging PRs from random people is probably not enjoyable. I am just trying to make sure the reason the PR has been stuck for a while is not because the code is not up to par. If that is the case I can definitely rework a lot of it as I did not start from scratch and the code might look like a Frankeinstein's monster. If the feature is still wanted but the PR is not adequate I would be glad to start again from scratch if it means the feature gets added.

Just let me know and thank you for all the good work you do on VSCodeVim

roosephu commented:
This PR makes sneak in VSCode-vim much more usable. Thank you, w-cantin for the great contribution! 

However, I observed an unexpected behavior with the `vscodevim-sneak-fork.vscodevim-sneak-fork` extension, which I suppose is built from w-cantin/Vim: sometimes `df` deletes one more character. I would like to open an issue in w-cantin/Vim but couldn't find the Issues tab. 

Settings: `"vim.sneakReplacesF": true,`
How to reproduce: The buffer has text `abcd` with the cursor on `a`. Then hit `dfb`.
Expected: The buffer becomes `cd`.
What happened: The buffer becomes `d`. 

Also tried `vscodevim.vim` and didn't find this issue. Disabling `vim.sneakReplacesF` also prevents it. Let me know if you need more details.

macintacos commented:
I don't mean to be "that guy" to chime in on an open PR, but I've been so pumped for this change to land that I can't help myself.

@J-Fields hope that this can be merged soon! This seems to be in a milestone for a version that has long-since been released - I'm wondering if perhaps this review has gotten lost in the shuffle?

roryfahy commented:
Would definitely love this feature
